File: src/stores/connection.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import type { ConnectionConfig, ConnectionStatus, StoredConnection } from '@/types/database'
import { connectionApi } from '@/api'
import { withErrorHandler } from '@/utils/errorHandler'

export const useConnectionStore = defineStore('connection', () => {
  // 状态
  const connections = ref<ConnectionConfig[]>([])
  const activeConnectionId = ref<string | null>(null)
  const connectionStatuses = ref<Map<string, ConnectionStatus>>(new Map())

  // 获取所有连接
  async function fetchConnections() {
    return withErrorHandler(async () => {
      connections.value = await connectionApi.getConnections()
    }, { messagePrefix: '获取连接列表失败' })
  }

  /**
   * 将 ConnectionConfig 转换为存储格式
   */
  function toStoredConnection(config: ConnectionConfig, isNew: boolean = false): StoredConnection {
    return {
      id: config.id,
      name: config.name,
      db_type: config.db_type,
      host: config.host,
      port: config.port,
      username: config.username,
      database: config.database,
      ssl: config.ssl,
      connection_timeout: config.connection_timeout,
      pool_size: config.pool_size,
      tags: config.tags || [],
      created_at: isNew ? Date.now() : config.created_at,
      updated_at: Date.now(),
    }
  }

  // 保存连接
  async function saveConnection(config: ConnectionConfig, password?: string) {
    return withErrorHandler(async () => {
      const storedConnection = toStoredConnection(config, true)
      const saved = await connectionApi.saveConnection(storedConnection, password || null)
      
      const index = connections.value.findIndex(c => c.id === saved.id)
      if (index >= 0) {
        connections.value[index] = { ...config, ...saved }
      } else {
        connections.value.push({ ...config, ...saved })
      }
      return saved
    }, { messagePrefix: '保存连接失败' })
  }

  // 更新连接
  async function updateConnection(config: ConnectionConfig, password?: string) {
    return withErrorHandler(async () => {
      const storedConnection = toStoredConnection(config, false)
      const updated = await connectionApi.updateConnection(storedConnection, password || null)
      
      const index = connections.value.findIndex(c => c.id === config.id)
      if (index >= 0) {
        connections.value[index] = { ...config, ...updated }
      }
      return updated
    }, { messagePrefix: '更新连接失败' })
  }

  // 删除连接
  async function deleteConnection(id: string) {
    return withErrorHandler(async () => {
      await connectionApi.deleteConnection(id)
      connections.value = connections.value.filter(c => c.id !== id)
      if (activeConnectionId.value === id) {
        activeConnectionId.value = null
      }
      connectionStatuses.value.delete(id)
    }, { messagePrefix: '删除连接失败' })
  }

  // 测试连接
  async function testConnection(config: ConnectionConfig) {
    return withErrorHandler(async () => {
      const result = await connectionApi.testConnection(config)
      if (!result.success) {
        throw new Error(result.message || '连接失败')
      }
      return result
    }, { messagePrefix: '测试连接失败' })
  }

  // 连接到数据库
  async function connectToDatabase(id: string) {
    const conn = connections.value.find(c => c.id === id)
    if (!conn) {
      throw new Error('连接配置不存在')
    }

    return withErrorHandler(async () => {
      await connectionApi.createConnection(id)
      updateConnectionStatus(id, 'connected')
    }, { 
      messagePrefix: '连接数据库失败',
      onError: () => updateConnectionStatus(id, 'error')
    })
  }

  // 断开数据库连接
  async function disconnectFromDatabase(id: string) {
    return withErrorHandler(async () => {
      await connectionApi.disconnectDatabase(id)
      updateConnectionStatus(id, 'disconnected')
    }, { messagePrefix: '断开连接失败' })
  }

  // 设置活动连接
  function setActiveConnection(id: string | null) {
    activeConnectionId.value = id
  }

  // 更新连接状态
  function updateConnectionStatus(id: string, status: ConnectionStatus) {
    connectionStatuses.value.set(id, status)
  }

  // 获取连接状态
  function getConnectionStatus(id: string): ConnectionStatus {
    return connectionStatuses.value.get(id) || 'disconnected'
  }

  // 获取活动连接
  function getActiveConnection(): ConnectionConfig | null {
    if (!activeConnectionId.value) return null
    return connections.value.find(c => c.id === activeConnectionId.value) || null
  }

  return {
    connections,
    activeConnectionId,
    connectionStatuses,
    fetchConnections,
    saveConnection,
    updateConnection,
    deleteConnection,
    testConnection,
    connectToDatabase,
    disconnectFromDatabase,
    setActiveConnection,
    updateConnectionStatus,
    getConnectionStatus,
    getActiveConnection,
  }
})

File: src-tauri/src/commands/connection.rs
use crate::database::{ConnectionConfig, DatabaseType, sqlite::SqliteDatabase};
use crate::models::{ConnectionTestResult, StoredConnection};
use crate::utils::crypto;
use crate::AppState;
use serde_json::{json, Value};
use tauri::{AppHandle, State};
use tauri_plugin_store::StoreExt;
use tracing::{info, instrument};

/// 将 StoredConnection 转换为 ConnectionConfig
fn stored_to_config(stored: &StoredConnection) -> Result<ConnectionConfig, String> {
    // 解密密码
    let password = if let Some(ref encrypted) = stored.encrypted_password {
        crypto::decrypt_password(encrypted)?
    } else {
        String::new()
    };
    
    Ok(stored_to_config_with_password(stored, &password))
}

/// 将 StoredConnection 转换为 ConnectionConfig（带明文密码）
fn stored_to_config_with_password(stored: &StoredConnection, password: &str) -> ConnectionConfig {
    let db_type = match stored.db_type.as_str() {
        "mysql" => DatabaseType::MySQL,
        "postgresql" => DatabaseType::PostgreSQL,
        "sqlite" => DatabaseType::SQLite,
        "mongodb" => DatabaseType::MongoDB,
        "redis" => DatabaseType::Redis,
        _ => DatabaseType::MySQL,
    };

    ConnectionConfig {
        id: stored.id.clone(),
        name: stored.name.clone(),
        db_type,
        host: stored.host.clone(),
        port: stored.port,
        username: stored.username.clone(),
        password: password.to_string(),
        database: stored.database.clone(),
        ssl: stored.ssl,
        connection_timeout: stored.connection_timeout,
        pool_size: stored.pool_size,
    }
}

/// 创建新的 SQLite 数据库文件
#[tauri::command]
#[instrument]
pub async fn create_sqlite_database(path: String) -> Result<String, String> {
    info!(path = %path, "收到创建 SQLite 数据库请求");
    SqliteDatabase::create_database_file(&path).map_err(|e| e.to_string())?;
    Ok("数据库创建成功".to_string())
}

/// 测试数据库连接
#[tauri::command]
#[instrument(skip(state, config))]
pub async fn test_connection(
    config: Value,
    state: State<'_, AppState>,
) -> Result<ConnectionTestResult, String> {
    let start = std::time::Instant::now();

    // 解析配置
    let conn_config: ConnectionConfig =
        serde_json::from_value(config).map_err(|e| e.to_string())?;

    // 使用连接管理器测试连接
    let manager = state.connection_manager.lock().await;
    let result = manager.test_connection(&conn_config).await;

    match result {
        Ok(_) => Ok(ConnectionTestResult {
            success: true,
            message: "连接成功".to_string(),
            version: None,
            ping_time_ms: start.elapsed().as_millis(),
        }),
        Err(e) => Ok(ConnectionTestResult {
            success: false,
            message: e.to_string(),
            version: None,
            ping_time_ms: start.elapsed().as_millis(),
        }),
    }
}

/// 保存连接配置
#[tauri::command]
pub async fn save_connection(
    app: AppHandle,
    mut connection: StoredConnection,
    password: Option<String>,
) -> Result<StoredConnection, String> {
    let store = app
        .store("connections.json")
        .map_err(|e| format!("Failed to get store: {}", e))?;

    // 如果提供了密码，加密并保存
    if let Some(pwd) = password {
        if !pwd.is_empty() {
            connection.encrypted_password = Some(crypto::encrypt_password(&pwd)?);
        }
    }

    // 保存连接信息
    store.set(connection.id.clone(), json!(connection));
    
    store.save().map_err(|e| e.to_string())?;

    Ok(connection)
}

/// 更新连接配置
#[tauri::command]
pub async fn update_connection(
    app: AppHandle,
    mut connection: StoredConnection,
    password: Option<String>,
) -> Result<StoredConnection, String> {
    let store = app
        .store("connections.json")
        .map_err(|e| format!("Failed to get store: {}", e))?;

    // 检查连接是否存在
    if !store.has(connection.id.clone()) {
        return Err("连接配置不存在".to_string());
    }

    // 如果提供了密码，加密并保存
    if let Some(pwd) = password {
        if !pwd.is_empty() {
            connection.encrypted_password = Some(crypto::encrypt_password(&pwd)?);
        }
    } else {
        // 如果没有提供新密码，保留原有密码
        if let Some(existing_value) = store.get(connection.id.clone()) {
            if let Ok(existing_conn) = serde_json::from_value::<StoredConnection>(existing_value) {
                connection.encrypted_password = existing_conn.encrypted_password;
            }
        }
    }

    // 更新连接信息
    store.set(connection.id.clone(), json!(connection));
    
    store.save().map_err(|e| e.to_string())?;

    Ok(connection)
}

/// 获取所有连接配置
#[tauri::command]
pub async fn get_connections(app: AppHandle) -> Result<Vec<StoredConnection>, String> {
    let store = app
        .store("connections.json")
        .map_err(|e| format!("Failed to get store: {}", e))?;

    let mut connections = Vec::new();

    // 遍历所有键值对
    for (_, value) in store.entries() {
        if let Ok(conn) = serde_json::from_value::<StoredConnection>(value.clone()) {
            connections.push(conn);
        }
    }

    Ok(connections)
}

/// 删除连接配置
#[tauri::command]
pub async fn delete_connection(app: AppHandle, id: String) -> Result<bool, String> {
    let store = app
        .store("connections.json")
        .map_err(|e| format!("Failed to get store: {}", e))?;

    store.delete(id);
    store.save().map_err(|e| e.to_string())?;

    Ok(true)
}

/// 创建数据库连接
#[tauri::command]
#[instrument(skip(state, app))]
pub async fn create_connection(
    app: AppHandle,
    connection_id: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    // 从存储中获取连接配置
    let store = app
        .store("connections.json")
        .map_err(|e| format!("Failed to get store: {}", e))?;
    
    let stored_value = store
        .get(connection_id.clone())
        .ok_or("连接配置不存在")?;
    
    let stored_conn: StoredConnection = serde_json::from_value(stored_value)
        .map_err(|e| format!("解析连接配置失败: {}", e))?;
    
    // 转换为 ConnectionConfig
    let config = stored_to_config(&stored_conn)?;
    
    // 使用连接管理器创建连接
    let manager = state.connection_manager.lock().await;
    manager.create_connection(config).await.map_err(|e| e.to_string())?;
    
    Ok(())
}

/// 断开数据库连接
#[tauri::command]
#[instrument(skip(state))]
pub async fn disconnect_database(
    connection_id: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    // 使用连接管理器断开连接
    let manager = state.connection_manager.lock().await;
    manager.disconnect(&connection_id).await.map_err(|e| e.to_string())?;
    
    Ok(())
}
